packages/blade/src/components/BottomSheet/BottomSheet.tsx

```diff
@@ -0,0 +1,413 @@
+    // @ts-expect-error this is a mutable ref
+    originalFocusElement.current = null;
+  }, [originalFocusElement]);
+
+  const focusOnInitialRef = React.useCallback(() => {
+    if (!initialFocusRef) {
+      // focus on close button
+      defaultInitialFocusRef.current?.focus();
+    } else {
+      // focus on the initialRef passed by the user
+      initialFocusRef.current?.focus();
+    }
+  }, [initialFocusRef]);
+
+  const close = React.useCallback(() => {
+    setPosY(0);
+    returnFocus();
+    onDismiss?.();
+    // close the select dropdown as well
+    dropdownBottomSheetProps?.setIsOpen(false);
+    scrollLockRef.current.deactivate();
+  }, [setPosY, returnFocus, dropdownBottomSheetProps, scrollLockRef, onDismiss]);
+
+  const open = React.useCallback(() => {
+    setPosY(dimensions.height * 0.5);
+    scrollLockRef.current.activate();
+    // @ts-expect-error this is a mutable ref
+    originalFocusElement.current = originalFocusElement.current ?? document.activeElement;
+    focusOnInitialRef();
+  }, [dimensions.height, focusOnInitialRef, scrollLockRef, setPosY]);
+
+  // sync controlled state to our actions
+  React.useEffect(() => {
+    if (isOpen === true) {
+      open();
+    }
+    if (isOpen === false) {
+      close();
+    }
+  }, [close, initialFocusRef, isOpen, open]);
+
+  // sync the select dropdown's state with bottomsheet's state
+  React.useEffect(() => {
+    if (!dropdownBottomSheetProps) return;
+
+    // this will let the Dropdown component know that it's rendering a bottomsheet
+    dropdownBottomSheetProps.setHasBottomSheet(true);
+
+    if (dropdownBottomSheetProps.isOpen) {
+      open();
+    }
+
+    if (!dropdownBottomSheetProps.isOpen && dropdownBottomSheetProps.selectionType === 'single') {
+      close();
+    }
+  }, [close, open, dropdownBottomSheetProps]);
+
+  /*
+      1. The content should not be scrollable on lower or middle snapPoints
+      2. If we reach the top snapPoint we make the content scrollable
+      3. scrolling down the content will work as usual
+      4. but if the scroll position is at top and then we drag down on the content body
+         the bottom-sheet will start the dragging and we will set the scroll to 'none'
+    */
+  const bind = useDrag(
+    ({
+      active,
+      last,
+      cancel,
+      tap,
+      movement: [_mx, my],
+      velocity: [_vx, vy],
+      lastOffset: [_, lastOffsetY],
+      down,
+      args: [{ isContentDragging = false } = {}] = [],
+    }) => {
+      setIsDragging(down);
+      const rawY = lastOffsetY - my;
+
+      const lowerSnapPoint = dimensions.height * snapPoints[0];
+      const upperSnapPoint = dimensions.height * snapPoints[snapPoints.length - 1];
+
+      // predictedY is used to create velocity driven swipe
+      // the faster you swipe the more distance you cover
+      // this enables users to reach upper & lower snappoint with a single swipe
+      const predictedDistance = my * (vy / 2);
```

Comment:
Did we consider using spring animation here?

Reply:
The rubberbandIfOutOfBounds function from useGesture kinda works like springs but in bottomsheet as a whole doesn't use any spring animations.